[TOPIC-GPIO] drivers: gpio_esp32: update to use new GPIO API

ExtremeGTX opened the pull request:
- Updates gpio driver and device tree files to the new GPIO Config flags
- Implements the new port_* APIs
- Fixes GPIO Bank1 config
- Add esp32.overlay to gpio_basic_api test

Tests:
- samples/basic/blinky
- samples/basic/button
- tests/drivers/gpio/gpio_basic_api
- tests/drivers/gpio/gpio_api_1pin

Board:
- esp32 DevKitC V4


Known issues:
part of `callback_variants` tests are failing
```
GPIO_INT_LEVEL_HIGH/LOW,
GPIO_INT_LEVEL_ACTIVE/INACTIVE
GPIO_INT_EDGE_BOTH
```

will continue working on it and update this PR

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

pabigot commented on drivers/gpio/gpio_esp32.c:
```diff
@@ -83,37 +78,11 @@ static inline u32_t *gpio_pin_reg(int pin)
 	volatile u32_t *reg = (u32_t *)(GPIO_FUNC0_IN_SEL_CFG_REG + pin * 4U);
 
-	if (flags & GPIO_POL_INV) {
+	if (flags & GPIO_ACTIVE_LOW) {
```
I wonder if this might be the problem with the test failures.  The wrapper functions in `gpio.h` now take responsibility for handling polarity by inverting various flags and values so that the driver sees everything as though it were active high.  If you're enabling hardware polarity inversion in some circumstances the results will be wrong.

ExtremeGTX replied:
as we discussed on slack, i removed `config_polarity` function.

mnkp commented on drivers/gpio/gpio_esp32.c:
```diff
@@ -137,37 +106,44 @@ static int gpio_esp32_config(struct device *dev, int access_op,
+		*data->port.write.set_reg = BIT(pin);
+	} else if (flags & GPIO_OUTPUT_INIT_LOW) {
+		*data->port.write.clear_reg = BIT(pin);
+	}
```
Initial pin value should be set before we configure the pin in the output mode. Otherwise we may see a glitch.

mnkp commented on drivers/gpio/gpio_esp32.c:
```diff
@@ -124,7 +93,7 @@ static void config_drive_strength(u32_t pin, int flags)
 	volatile u32_t *reg = gpio_pin_reg(pin);
 
-	if ((flags & GPIO_DS_DISCONNECT_LOW) == GPIO_DS_DISCONNECT_LOW) {
+	if ((flags & GPIO_OPEN_SOURCE) == GPIO_OPEN_SOURCE) {
```
According to the reference manual ESP32 platform supports GPIO open drain mode, not open source. This section should be written as
```
if (flags & GPIO_SINGLE_ENDED) {
	if (flags & GPIO_LINE_OPEN_DRAIN) {
		*reg |= GPIO_PIN_PAD_DRIVER;
	} else {
		r = -ENOTSUP;
	}
} else {
	*reg &= ~GPIO_PIN_PAD_DRIVER;
}
```
and placed in `gpio_esp32_config` function.

ExtremeGTX commented on drivers/gpio/gpio_esp32.c:
```diff
@@ -200,12 +176,101 @@ static int gpio_esp32_read(struct device *dev, int access_op,
+	/* Bit 3 of INT_ENA will enable interrupts on CPU 0 */
+	v |= (1 << 2) << GPIO_PIN_INT_ENA_S;
+	/* Interrupt triggering mode */
+	v |= type << GPIO_PIN_INT_TYPE_S;
```
- Bit 3 is TRM mistake, it should be Bit 2.
- updated in the new commit

ExtremeGTX commented:
tests/drivers/gpio/gpio_basic_api

The following tests still fails:
```
GPIO_INT_LEVEL_HIGH/LOW,
GPIO_INT_LEVEL_ACTIVE/INACTIVE
GPIO_INT_EDGE_BOTH
```
`GPIO_INT_EDGE_BOTH` is working in application code but test case is not working, i tested it on DevKit V4 and M5-Stack using GPIOs and using Buttons.

The problem with LEVEL interrupts is clearing the interrupt, dunno why clearing status register is not working.
it is only working in esp-idf test suite

mnkp commented on drivers/gpio/gpio_esp32.c:
```diff
@@ -355,6 +428,9 @@ GPIO_DEVICE_INIT(DT_INST_1_ESPRESSIF_ESP32_GPIO_LABEL, 32_to_39);
 {
+	/* FIXME: There is a problem in clearing interrupts
+	 *  of type LEVEL_LOW/HIGH, hence it is not usable
+	 */
```
We should move this code from `gpio_esp32_fire_callbacks` here
```
	struct gpio_esp32_data *data = device->driver_data;
	u32_t values = *data->port.irq.status_reg;

	*data->port.irq.ack_reg = values;
```
It conceptually belongs to ISR routine, also we need to write to `ack_reg` before executing callbacks.

mnkp commented on drivers/gpio/gpio_esp32.c:
```diff
@@ -355,6 +428,9 @@ GPIO_DEVICE_INIT(DT_INST_1_ESPRESSIF_ESP32_GPIO_LABEL, 32_to_39);
 static void gpio_esp32_isr(void *param)
 {
+	/* FIXME: There is a problem in clearing interrupts
+	 *  of type LEVEL_LOW/HIGH, hence it is not usable
```
We shouldn't go too deeply into the details of the issue, however the fact that the level interrupts cannot be cleared is the expected behavior. The unexpected behavior is that they are not re-triggered, despite that the interrupt status bit set high the interrupt routine is not invoked.

ExtremeGTX replied:
> The unexpected behavior is that they are not re-triggered, despite that the interrupt status bit set high the interrupt routine is not invoked.

The problem is the new interrupt won't be triggered until the status flag is cleared, and  this clearing is not done properly!.

Here are what i tried to do:
- Write 0x0000 0000 to STATUS_REG directly
- Write 0xFFFF FFFF to W1TC_REG
- Write 0xFFFF FFFF to W1TC_REG using assembly instructions
- shorten the ISR to just clear the STATUS flag (no function calls at all).

Other thing i noticed in another implementations is, all of them include **a loop**  to check and clear interrupt status for each GPIO. this looks weird to me because ISRs should be in general small, fast and for sure no loops at all.

I don't know if @andyross faced something similar while adding support for xtensa architecture.

mnkp reviewed (changes requested):
When the issues are fixed, could you also try to run `tests/drivers/gpio/gpio_api_1pin/` testcase. It will use the pin where LED0 is connected to perform the test. The test requires pin to work in simultaneous input/output mode. According to the documentation and comments in the `pinmux_esp32.c` file this is supported.

pabigot commented:
I think I've resolved the core problem; at least in a way that makes sense to me: you can't support both edge and level interrupts simultaneously on this hardware.

`CONFIG_GPIO_ESP32_IRQ` defaults to 10, which is an edge-triggered interrupt.  The test currently fails on the level-triggered tests.

If I set `CONFIG_GPIO_ESP32_IRQ=12` in `prj.conf` then the level tests pass and the edge-triggered ones fail.

There are no CPU interrupts that are listed as both edge and level triggered.

Now it's possible all that is wrong, and in fact you can use any CPU interrupt line for both edge and level interrupts, but it doesn't look that way to me.

So in https://github.com/pabigot/zephyr/commits/pr/19753 I've added a commit that detects the supported trigger type of the configured GPIO interrupt line and rejects attempts to configure for unsupported interrupt types.  I also tweaked some macros/comments, and moved the interrupt acknowledge before the callback.  I have verified that this does in fact clear the status register, so the fact that it didn't may be because the level configuration conflicted with the interrupt support.

@ExtremeGTX you are welcome to cherry-pick and squash that into your work.  This passes the two-pin test with line 10 (edge) and line 12 (level) (or will once #19784 gets merged).

However: I cannot make the both-edge case work, so I've just disabled support for that.

mnkp commented:
> I think I've resolved the core problem; at least in a way that makes sense to me: you can't support both edge and level interrupts simultaneously on this hardware.
> 
> There are no CPU interrupts that are listed as both edge and level triggered.

Well spotted, that indeed seems to be the culprit. It is possible to attach more than one "Peripheral Interrupt Source" to the same CPU IRQ, e.g. we can have SPI and I2C modules trigger interrupt on the same IRQ line, however it doesn't seem possible to attach the same "Peripheral Interrupt Source" to two different CPU IRQ lines. In this case, the GPIO module can indeed respond either to edge or level interrupts but not both since any given IRQ line is predefined as either edge or level. 

That complicates the driver. One clean way to handle this situation would be to add a function which would take IRQ line number as a parameter and return the type of the interrupt this line supports: edge, level or invalid (since some IRQs are internal to the CPU). Then the driver would need to return `-ENOTSUP` for the type of the interrupt which is not supported.

pabigot commented:
> One clean way to handle this situation would be to add a function which would take IRQ line number as a parameter and return the type of the interrupt this line supports: edge, level or invalid (since some IRQs are internal to the CPU). Then the driver would need to return `-ENOTSUP` for the type of the interrupt which is not supported.

Like [this](https://github.com/pabigot/zephyr/commit/9aa73dc41cb914ee8d72e832d0a4f51784491156), perhaps.

mnkp commented:
> Like [this](https://github.com/pabigot/zephyr/commit/9aa73dc41cb914ee8d72e832d0a4f51784491156), perhaps.

Looks good. But maybe we should put `switch (trig) {` in the `else` branch. The macro verifying the supported interrupt type gets resolved at the compilation stage. With the `else` the `switch (trig) {` would only get compiled in if supported.

ExtremeGTX commented:
@pabigot Good Catch 👍
Thanks for this, but dunno why you disabled BOTH_EDGE interrupts, it works in application (i know testcase is not passing), but i tested it with an ACTIVE LOW Button and i got 2 interrupts.
One on button press and the other on button release.
did i miss anything here ?

pabigot commented:
> Looks good. But maybe we should put `switch (trig) {` in the `else` branch. The macro verifying the supported interrupt type gets resolved at the compilation stage. With the `else` the `switch (trig) {` would only get compiled in if supported.

Without the `else` the compiler will discard the unreachable code that would follow the unconditional `return`.

> Thanks for this, but dunno why you disabled BOTH_EDGE interrupts, it works in application (i know testcase is not passing),

That'd be why.  There may be circumstances where it works, but it's not reliable.

carlescufi commented:
@ExtremeGTX can you cherry-pick @pabigot's commit and squash it into yours and then the PR will be ready for merging.

ExtremeGTX commented:
@carlescufi sure, will do it tomorrow morning.
But, am still want to leave EDGE_BOTH mode enabled, since it works with normal applications.

In the end, it is up to you

pabigot commented:
@ExtremeGTX The general position when discussed today was that if the test doesn't pass we can't enable it, because something's clearly not right.  We can continue to investigate and enable it later if we can figure out what's going on.  I may find some time for that, or you can look at it, but we'd really like to get the basic driver in ASAP.

ExtremeGTX commented:
ok, i will update this PR tomorrow morning

mnkp commented:
> `GPIO_INT_EDGE_BOTH` is working in application code but test case is not working, i tested it on DevKit V4 and M5-Stack using GPIOs and using Buttons.

I don't know the evaluation board in question. However - unless there is some extra circuitry - signal from the button will bounce / won't be clean. It's not a reliable way to verify edge interrupts. Having passing testcase would be preferable.

ExtremeGTX commented:
The new push includes all changes requested by @mnkp and changes done by @pabigot 
dunno if i should add something in commit msg to give credits for @pabigot work.

pabigot reviewed (approved):
Code looks good.  No need for acknowledgment for such a small change.

The commit message is a little unusual, but except for the subject line including `[TOPIC-GPIO]` I don't think it violates any standards.   There are existing merged PRs that have that same prefix though, so it can be fixed in a cleanup pass over the branch after it's merged.

mnkp reviewed (commented):
LGTM with minor comment.

ExtremeGTX committed:
[TOPIC-GPIO] drivers: gpio_esp32: update to use new GPIO API

- Updates gpio driver and device tree files to the new GPIO Config flags
- Implements the new port_* APIs
- Update I2C and PWM Drivers to use new GPIO config
- Add esp32.overlay to gpio_basic_api test
- refactor convert_int_type, regs struct
- remove config_polarity
- add kConfig notes

Tests:
- samples/basic/blinky
- samples/basic/button
- tests/drivers/gpio/gpio_basic_api
- tests/drivers/gpio/gpio_api_1pin

Board:
- esp32 DevKitC V4

Note about interrupts:
The ESP32 requires specifying a CPU interrupt to be used for GPIO
interrupt signals.  CPU interrupts can be either level or edge (or
special) triggered, but not both.
Please check gpio/Kconfig.esp32 for more info.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX commented on drivers/gpio/gpio_esp32.c:
```diff
@@ -20,188 +20,270 @@
+		volatile u32_t *input_reg;
+		volatile u32_t *output_reg;
+		volatile u32_t *status_reg;
+		volatile u32_t *ack_reg;
```
resolved
